File: src/main/java/model/util/Position.java
package model.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.Objects;

/**
 * Une position du boardregion.
 * (0,0) est en haut à gauche
 */
public class Position implements Serializable {
    public static final int LIMITE = 8;

    private final int row;
    private final int column;

    public Position(int row, int column) {
        this.row = row;
        this.column = column;
    }

    public int getColumn() {
        return column;
    }

    public int getRow() {
        return row;
    }

    /**
     * Retourne une position déplacé
     *
     * @return la nouvelle position
     */
    @NotNull
    public Position shift(Offset offset) {
        return new Position(row + offset.getVerticalShift(), column + offset.getHorizontalShift());
    }

    /**
     * @return si la position rentre dans les limites
     */
    public boolean isValid() {
        return 0 <= row && row < LIMITE && 0 <= column && column < LIMITE;
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        if (obj == null) return false;
        if (!(obj instanceof Position)) return false;
        if (((Position) obj).getRow() != this.row) return false;
        return ((Position) obj).getColumn() == this.column;
    }

    @Override
    public String toString() {
        return "r" + row + "c" + column;
    }


    @Override
    public int hashCode() {
        return Objects.hash(column, row);
    }
}

File: src/main/java/model/util/Position.kt
package model.util

import java.io.Serializable
import java.util.*

/**
 * A position (square) on the board. Top-left is (0,0)
 */
class Position(val row: Int, val column: Int) : Serializable {

    /**
     * @return if the position is within the limits (is a valid position on the board)
     */
    val isValid: Boolean
        get() = row in 0..(LIMIT - 1) && 0 <= column && column < LIMIT

    /**
     * @param offset how much to shift this position
     * @return a new position that is shifted from this
     */
    fun shift(offset: Offset): Position {
        return Position(row + offset.verticalShift, column + offset.horizontalShift)
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other !is Position) return false
        return if (other.row != this.row) false else other.column == this.column
    }

    override fun toString(): String {
        return "r" + row + "c" + column
    }


    override fun hashCode(): Int {
        return Objects.hash(column, row)
    }

    companion object {
        const val LIMIT = 8
    }
}
